### DISCUSSION TITLE
How do I use this in more complex scenarios?

### DISCUSSION THREAD

**Discussion by samgregson** [zou-group/textgrad#40] [Q&A]
The examples are somewhat lacking. I would love to use TextGrad for prompts which have a combination of non-optimisable context as well as optimisable text (like RAG). I'd also like to see examples of chains of LLM calls (not just a single LLM as per examples). Finally, the paper mentions optimisation for tool calling, how do we achieve this? Thanks for the awesome development, just wish I new how to use it in more complex settings!

**Comment by belmarca** [comment]
The examples do need a bit of love. E.g., in the Prompt Optimization tutorial notebook `run_validation_revert` has a `val_set` parameter which it passes to `eval_dataset`. However `eval_dataset` uses the name `test_set` for this variable.
